airflow/models/dagcode.py

```diff
@@ -0,0 +1,147 @@
+    This feature is controlled by:
+
+    * ``[core] store_serialized_dags = True``: enable this feature
+    * ``[core] store_dag_code = True``: enable this feature
+
+    For details on dag serialization see SerializedDagModel
+    """
+    __tablename__ = 'dag_code'
+
+    fileloc_hash = Column(
+        BigInteger, nullable=False, primary_key=True, autoincrement=False)
+    fileloc = Column(String(2000), nullable=False)
+    # The max length of fileloc exceeds the limit of indexing.
+    last_updated = Column(UtcDateTime, nullable=False)
+    source_code = Column(UnicodeText(), nullable=False)
+
+    def __init__(self, full_filepath: str):
+        self.fileloc = full_filepath
+        self.fileloc_hash = DagCode.dag_fileloc_hash(self.fileloc)
+        self.last_updated = timezone.utcnow()
+        self.source_code = DagCode._read_code(self.fileloc)
+
+    @classmethod
+    def _read_code(cls, fileloc: str):
+        with open_maybe_zipped(fileloc, 'r') as source:
+            source_code = source.read()
+        return source_code
+
+    @provide_session
+    def sync_to_db(self, session=None):
+        """Writes code into database.
+
+        :param session: ORM Session
+        """
+        old_version = session.query(
+            DagCode.fileloc, DagCode.fileloc_hash, DagCode.last_updated) \
+            .filter(DagCode.fileloc_hash == self.fileloc_hash) \
+            .first()
+
+        if old_version and old_version.fileloc != self.fileloc:
+            raise AirflowException(
+                "Filename '{}' causes a hash collision in the database with "
+                "'{}'. Please rename the file.".format(
+                    self.fileloc, old_version.fileloc))
+
+        file_modified = datetime.fromtimestamp(
+            os.path.getmtime(correct_maybe_zipped(self.fileloc)), tz=timezone.utc)
+
+        if old_version and (file_modified - timedelta(seconds=120)) < \
+                old_version.last_updated:
+            return
+
+        session.merge(self)
+
+    @classmethod
+    @provide_session
+    def bulk_sync_to_db(cls, filelocs: Iterable[str], session=None):
+        """Writes code in bulk into database.
+
+        :param filelocs: file paths of DAGs to sync
+        :param session: ORM Session
+        """
+        filelocs = set(filelocs)
+        for file in filelocs:
+            DagCode(file).sync_to_db(session=session)
+
+    @classmethod
+    @provide_session
+    def remove_deleted_code(cls, alive_dag_filelocs: List[str], session=None):
+        """Deletes code not included in alive_dag_filelocs.
+
+        :param alive_dag_filelocs: file paths of alive DAGs
+        :param session: ORM Session
+        """
+        alive_fileloc_hashes = [
+            cls.dag_fileloc_hash(fileloc) for fileloc in alive_dag_filelocs]
+
+        log.debug("Deleting code from %s table ", cls.__tablename__)
+
+        session.execute(
+            cls.__table__.delete().where(
```

Comment:
```suggestion
            session.query(cls).filter(
```
This way you don't need to use magic methods.

Reply:
I have changed it to what you suggested but SO seems to suggest previous solution was faster.
Here, I will first do query and then execute delete where id in ()
And with previous solution it would have been delete from where
https://stackoverflow.com/questions/39773560/sqlalchemy-how-do-you-delete-multiple-rows-without-querying